src/cargo/util/cache_lock.rs

```diff
@@ -0,0 +1,463 @@
+}
+
+/// Macro for acquiring a shared lock without blocking.
+macro_rules! do_try_shared_lock {
+    ($self: ident, $config: expr, $to_set: ident, $lock_name: expr, $unused: expr) => {
+        $self.$to_set = match $config.home().try_open_shared_create($lock_name) {
+            Ok(Some(lock)) => Some(lock),
+            Ok(None) => {
+                return Ok(false);
+            }
+            Err(e) => {
+                tracing::warn!("failed to acquire cache lock {}: {e:?}", $lock_name);
+                None
+            }
+        };
+    };
+}
+
+/// Macro for acquiring an exclusive lock (can block).
+macro_rules! do_exclusive_lock {
+    ($self: ident, $config: expr, $to_set: ident, $lock_name: expr, $descr: expr) => {
+        match $config.home().open_rw($lock_name, $config, $descr) {
+            Ok(lock) => {
+                $self.$to_set = Some(lock);
+                Ok(())
+            }
+            Err(e) => {
+                if maybe_readonly(&e) {
+                    // This is a best-effort attempt to at least try to
+                    // acquire some sort of lock. This can help in the
+                    // situation where this cargo only has read-only access,
+                    // but maybe some other cargo has read-write. This will at
+                    // least attempt to coordinate with it.
+                    //
+                    // We don't want to fail on a read-only mount because
+                    // cargo grabs an exclusive lock in situations where it
+                    // may only be reading from the package cache. In that
+                    // case, cargo isn't writing anything, and we don't want
+                    // to fail on that.
+                    do_shared_lock!($self, $config, $to_set, $lock_name, $descr);
+                    Ok(())
+                } else {
+                    Err(e).with_context(|| "failed to acquire package cache lock")
+                }
+            }
+        }
+    };
+}
+
+/// Macro for acquiring an exclusive lock without blocking.
+macro_rules! do_try_exclusive_lock {
+    ($self: ident, $config: expr, $to_set: ident, $lock_name: expr, $unused: expr) => {
+        match $config.home().try_open_rw($lock_name) {
+            Ok(Some(lock)) => {
+                $self.$to_set = Some(lock);
+                Ok(())
+            }
+            Ok(None) => return Ok(false),
+            Err(e) => {
+                if maybe_readonly(&e) {
+                    do_try_shared_lock!($self, $config, $to_set, $lock_name, $unused);
+                    Ok(())
+                } else {
+                    Err(e).with_context(|| "failed to acquire package cache lock")
+                }
+            }
+        }
+    };
+}
+
+/// Macro to help with acquiring a lock.
+///
+/// `shared` and `exclusive` are inputs to the macro so that either the
+/// blocking or non-blocking implementations can be called based on what the
+/// caller wants.
+macro_rules! do_lock {
```

Comment:
Why all the macros?

Reply:
Without the macros, the equivalent permutations would all need to be expanded as separate methods (in order to access the correct fields). It would maybe roughly double the code, duplicating parts that I think are quite subtle and important not to duplicate.

Reply:
I feel like at least some of them would be just as clear or clearer as functions that take `&mut` arguments, like `decrement` or `do_shared_lock`.

Reply:
I have rewritten this to not use macros. It's a little longer to accommodate some extra types, but I think still retains the goal of avoiding too much duplication.

Reply:
Thanks!